src/blob/index/bptree/core.rs

```diff
@@ -0,0 +1,287 @@
+            })
+            .map(|headers| (headers, self.header.records_count))
+    }
+
+    async fn get_any(&self, key: &[u8]) -> Result<Option<RecordHeader>> {
+        let root_offset = self.metadata.root_offset;
+        let mut buf = [0u8; BLOCK_SIZE];
+        let leaf_offset = self.find_leaf_node(key, root_offset, &mut buf).await?;
+        if let Some((first_header_offset, _amount)) =
+            self.find_first_header(leaf_offset, key, &mut buf).await?
+        {
+            let header = self.read_headers(first_header_offset, 1).await?.remove(0);
+            Ok(Some(header))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn validate(&self) -> Result<()> {
+        if self.header.written == 1 {
+            Ok(())
+        } else {
+            Err(anyhow!("Index Header is not valid"))
+        }
+    }
+}
+
+impl BPTreeFileIndex {
+    async fn find_leaf_node(&self, key: &[u8], mut offset: u64, buf: &mut [u8]) -> Result<u64> {
+        let key_size = key.len() as u64;
+        while offset >= self.metadata.tree_offset {
+            self.file.read_at(buf, offset).await?;
+            let node = Node::deserialize(buf, key_size)?;
+            offset = node.key_offset(&key);
+        }
+        Ok(offset)
+    }
+
+    async fn find_first_header(
+        &self,
+        leaf_offset: u64,
+        key: &[u8],
+        buf: &mut [u8],
+    ) -> Result<Option<(u64, u64)>> {
+        let leaf_size = key.len() + std::mem::size_of::<u64>();
+        let buf_size = self.leaf_node_buf_size(leaf_size, leaf_offset);
+        let header_pointers = self
+            .get_header_pointers(leaf_offset, key, buf, leaf_size, buf_size)
+            .await?;
+        self.search_header_pointer(header_pointers, key, buf_size as u64 + leaf_offset)
+    }
+
+    fn leaf_node_buf_size(&self, leaf_size: usize, leaf_offset: u64) -> usize {
+        let buf_size = (self.metadata.tree_offset - leaf_offset) as usize;
+        buf_size
+            .min(BLOCK_SIZE - (BLOCK_SIZE % leaf_size))
+            .min((self.metadata.tree_offset - leaf_offset) as usize)
+    }
+
+    fn search_header_pointer(
+        &self,
+        header_pointers: Vec<(Vec<u8>, u64)>,
+        key: &[u8],
+        absolute_buf_end: u64,
+    ) -> Result<Option<(u64, u64)>> {
+        match header_pointers.binary_search_by(|(cur_key, _)| cur_key.as_slice().cmp(key)) {
+            Err(_) => Ok(None),
+            Ok(i) => Ok(Some(self.with_amount(header_pointers, i, absolute_buf_end))),
+        }
+    }
+
+    async fn get_header_pointers(
+        &self,
+        leaf_offset: u64,
+        key: &[u8],
+        buf: &mut [u8],
+        leaf_size: usize,
+        buf_size: usize,
+    ) -> Result<Vec<(Vec<u8>, u64)>> {
+        let (buf, _rest) = buf.split_at_mut(buf_size as usize);
+        self.file.read_at(buf, leaf_offset).await?;
+        buf.chunks(leaf_size)
+            .map(|bytes| {
+                let (key, offset) = bytes.split_at(key.len());
```

Comment:
I am not sure about convenience, because in case of new `Leaf` struct I'll have to implement my own Serializer and Deserializer (because I don't need to store key vector size) + I need to pass key_size during serialization/deserialization, so I think this wrapper won't be more convenient imo